compiler/rustc_lint/src/types/improper_ctypes.rs

```diff
@@ -291,57 +346,133 @@ impl VisitorState {
+                }
+                if matches!(k, ty::RawPtr(..)) {
+                    ret |= Self::PTR_RAW;
+                }
+                ret
+            }
+            ty::Adt(..) => {
+                if ty.boxed_ty().is_some() {
+                    Self::IN_PTR | Self::PTR_MUT
+                } else {
+                    Self::IN_ADT | Self::MEMORY_INLINED
+                }
+            }
+            ty::Tuple(..) | ty::Array(..) | ty::Slice(_) => Self::MEMORY_INLINED,
+            k @ _ => bug!("Unexpected outer type {:?} of kind {:?}", ty, k),
+        }
+    }
+}
+
+impl VisitorState {
+    /// From an existing state, compute the state of any subtype of the current type.
+    fn get_next<'tcx>(&self, current_ty: Ty<'tcx>) -> Self {
+        debug_assert!(!matches!(current_ty.kind(), ty::FnPtr(..)));
+        Self {
+            persistent_flags: self.persistent_flags,
+            ephemeral_flags: EphemeralStateFlags::from_outer_ty(current_ty),
+        }
+    }
+    fn get_next_in_fnptr<'tcx>(&self, current_ty: Ty<'tcx>, is_ret: bool) -> Self {
+        debug_assert!(matches!(current_ty.kind(), ty::FnPtr(..)));
+        Self {
+            persistent_flags: if is_ret {
+                PersistentStateFlags::RETURN_TY_IN_FNPTR
+            } else {
+                PersistentStateFlags::ARGUMENT_TY_IN_FNPTR
+            },
+            ephemeral_flags: EphemeralStateFlags::from_outer_ty(current_ty),
+        }
+    }
+
+    /// Generate the state for an "outermost" type that needs to be checked
+    fn entry_point(persistent_flags: PersistentStateFlags) -> Self {
+        Self { persistent_flags, ephemeral_flags: EphemeralStateFlags::NO_OUTER_TY }
+    }
 
     /// Get the proper visitor state for a given function's arguments.
     fn argument_from_fnmode(fn_mode: CItemKind) -> Self {
-        match fn_mode {
-            CItemKind::Definition => VisitorState::ARGUMENT_TY_IN_DEFINITION,
-            CItemKind::Declaration => VisitorState::ARGUMENT_TY_IN_DECLARATION,
-        }
+        let p_flags = match fn_mode {
+            CItemKind::Definition => PersistentStateFlags::ARGUMENT_TY_IN_DEFINITION,
+            CItemKind::Declaration => PersistentStateFlags::ARGUMENT_TY_IN_DECLARATION,
+        };
+        Self::entry_point(p_flags)
     }
 
     /// Get the proper visitor state for a given function's return type.
     fn return_from_fnmode(fn_mode: CItemKind) -> Self {
-        match fn_mode {
-            CItemKind::Definition => VisitorState::RETURN_TY_IN_DEFINITION,
-            CItemKind::Declaration => VisitorState::RETURN_TY_IN_DECLARATION,
-        }
+        let p_flags = match fn_mode {
+            CItemKind::Definition => PersistentStateFlags::RETURN_TY_IN_DEFINITION,
+            CItemKind::Declaration => PersistentStateFlags::RETURN_TY_IN_DECLARATION,
+        };
+        Self::entry_point(p_flags)
+    }
+
+    /// Get the proper visitor state for a static variable's type
+    #[inline]
```

Comment:
Nit: no need to `#[inline]`, that will happen automatically on crate-private functions